plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java

```diff
@@ -86,6 +91,22 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
                 .orElse(null);
     }
 
+    @Override
+    public ConnectorTableProperties getTableProperties(ConnectorSession session, ConnectorTableHandle table)
+    {
+        JdbcTableHandle tableHandle = (JdbcTableHandle) table;
+        List<LocalProperty<ColumnHandle>> sortingProperties = tableHandle.getSortOrder()
```

Comment:
I think we could just do this in DefaultJdbcMetadata#getTableProperties
I don't see anything here unique to phoenix
Any other connector which can do TopN pushdown should automatically benefit in same way

Reply:
One case where this would not be true in the phoenix (4) connector. The topN is pushed in the plan, but the connection still can decide not honor it. The same is true for some of the other connector (Postgres, MySQL), which also also may decide not to do the topN *after* the plan has decided to push this down.

Reply:
Ok, doing it in DefaultJdbcMetadata just requires the assumption that the sortOrder in JdbcTableHandle will be guaranteed when the data is read from it, it's not required that the limit part of topN be guaranteed.
If that assumption also does not hold, then we can't move this to DefaultJdbcMetadata.
I'm not sure why we wouldn't drop the sortOrder from JdbcTableHandle if we weren't going to actually use that in forming the JDBC query.

Reply:
yes, in `DefaultJdbcMetadata` it would need to depend on `io.trino.plugin.jdbc.JdbcClient#isTopNLimitGuaranteed`

Reply:
@findepi is `isTopNLimitGuaranteed` just about whether the limit part of topN is guaranteed by connector or does it impact the sorting part as well ? 
If it's just about the limit, then I think don't need to worry about that, for this optimisation the main concern is whether the sorting part of TopN is guaranteed or not, we'll apply the limit part ourselves anyway.

Reply:
> s `isTopNLimitGuaranteed` just about whether the limit part of topN is guaranteed by connector or does it impact the sorting part as well ?

i am glad that you asked!

see https://github.com/trinodb/trino/pull/8238

Reply:
@findepi Thanks for clearing up isTopNLimitGuaranteed
I'm wondering what it would take to get this optimization into DefaultJdbcMetadata for cases where after TopN pushdown into TS, the result size is not strictly guaranteed (isTopNGuaranteed is false) but the ordering is.
Would a isTopNOrderingGuaranteed API in JdbcClient makes sense ? 
Why isn't the presence of sortOrder in JdbcTableHandle sufficient to say that the result will be ordered ? We are using sortOrder in applyQueryTransformations if it exists, so I'm not yet understanding why the output wouldn't be sorted even if TopN is not guaranteed.

Reply:
I think for now let's leave it in the Phoenix connector.

A special case is the Phoenix (4) connector, which overrides the topN function in some cases. That scenario would not work in general, right? As it would not even push the order by down.

Btw. order by pushdown (without limit) would be a useful feature anyway.

Reply:
As for isTopNGuaranteed... Phoenix will silently not enforce a limit when the limit > 2^31. Does that mean now that topN is guaranteed or not?

Reply:
@lhofhansl good point. If Phoenix 5 is most interesting to you & special case anyway, let's proceed with Phoenix, but let's define what we want to achieve in general, in an issue, so that's easier to follow up